### ISSUE TITLE
Can no longer connect to Panoply Redshift databases

### ISSUE THREAD

**Issue by patricktrainer** [metabase/metabase#11335]
**Describe the bug**
As of this morning, Metabase is no longer able to connect to our Redshift instance. 
After upgrading to 33.5.1 from 33.4.1 the issue still persists. 

**Logs**
These are the produced errors in the logs: 

```{:class com.mchange.v2.resourcepool.CannotAcquireResourceException,
  :error "A ResourcePool could not acquire a resource from its primary factory or source.",
  :cause
  {:class java.sql.SQLException,
   :error "[Amazon](600000) Error setting/closing connection: extension (5) should not be presented in certificate_request.",
   :cause
   {:class com.amazon.support.exceptions.GeneralException,
    :error "[Amazon](600000) Error setting/closing connection: extension (5) should not be presented in certificate_request.",
    :cause {:class javax.net.ssl.SSLHandshakeException, :error "extension (5) should not be presented in certificate_request"}}}}}
```

**Information about your Metabase Installation:**

{
  "browser-info": {
    "language": "en-US",
    "platform": "MacIntel",
    "userAgent": "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.14; rv:70.0) Gecko/20100101 Firefox/70.0",
    "vendor": ""
  },
  "system-info": {
    "java.runtime.name": "OpenJDK Runtime Environment",
    "java.runtime.version": "11.0.3+7",
    "java.vendor": "AdoptOpenJDK",
    "java.vendor.url": "https://adoptopenjdk.net/",
    "java.version": "11.0.3",
    "java.vm.name": "OpenJDK 64-Bit Server VM",
    "java.vm.version": "11.0.3+7",
    "os.name": "Linux",
    "os.version": "4.14.97-74.72.amzn1.x86_64",
    "user.language": "en",
    "user.timezone": "GMT"
  },
  "metabase-info": {
    "databases": [
      "redshift",
      "mysql",
      "h2"
    ],
    "hosting-env": "elastic-beanstalk",
    "application-database": "postgres",
    "run-mode": "prod",
    "version": {
      "date": "2019-11-13",
      "tag": "v0.33.5.1",
      "branch": "release-0.33.x",
      "hash": "26159d2"
    },
    "settings": {
      "report-timezone": "US/Central"
    }
  }
}

**Severity**
This is blocking my orgs use entirely.

**Comment by GitsMcGee** [comment]
As a second data point, this was happening to me as well this morning, but seems to have gone away on its own. Perhaps this is an error on the Redshift side of the equation? @patricktrainer is it still broken for you?

**Comment by mazameli** [comment]
> After upgrading to 33.5.1 from 33.4.1 the issue still persists.

To clarify, @patricktrainer, this was also happening to you on 0.33.4? @GitsMcGee, which version are you running?

**Comment by GitsMcGee** [comment]
@mazameli I'm running 0.33.5.

**Comment by mazameli** [comment]
Cool, thanks. It looks like our own redshift connection is working, so we'll need to do a bit of digging into this. I'll be very interested to know if this issue persists for Patrick.

**Comment by patricktrainer** [comment]
@mazameli yes, happens in both versions. 

And @GitsMcGee - yep, still not working. Fingers crossed that it sorts itself out...

**Comment by dlondero** [comment]
Having the same since yesterday on `33.5.1`. Downgrade didn't help. Connecting directly to RedShift using an external client works.

Most probably it is something connected with `elastic-beanstalk` where we have MB deployed on a Docker container.

**Comment by ppp0** [comment]
See https://stackoverflow.com/questions/52016415/jdk-11-ssl-error-on-valid-certificate-working-in-previous-versions
It seems to be on Redshift's side (Using Panoply's here).

Redshift JDBC Drivers tried:
- 1.2.36.1060
- 1.2.16.1027
- 1.2.10.1009
- 1.2.1001  <---- This one works

**Comment by patricktrainer** [comment]
I'm wondering if #11147 may be culprit? 

From @ppp0 's digging and the [AWS docs](https://docs.aws.amazon.com/redshift/latest/mgmt/jdbc-previous-versions.html), newer drivers may have compatibility issues.

> Amazon Redshift has changed the way that we manage SSL certificates. If you must use a driver version earlier than 1.2.8.1005, you might need to update your current trust root CA certificates to continue to connect to your clusters using SSL.

**Comment by ppp0** [comment]
If I run Metabase on a my computer's Docker, I do not have these issues, btw. - Works out of the box with the provided Redshift JDBC driver

**Comment by ifosch** [comment]
This happened to us as well. We were using 0.32.10. Tried upgrading to 0.33.5.1, without success. Downgrading to 0.32.9 made this work fine again. I am guessing this could be related to [AWS RDS SSL certificates change](https://forums.aws.amazon.com/thread.jspa?messageID=920130), even while we got the problem running against an existing Redshift node, and combining with some JDBC configuration... but I'm just guessing here.

**Comment by ppp0** [comment]
I can confirm @ifosch 's findings: downgrading to `v0.32.9` fixed it

**Comment by GitsMcGee** [comment]
Scratch what I said earlier. It's down again for me. I'm on 0.33.5 and running on Elastic Beanstalk.

**Comment by flamber** [comment]
It's interesting if 0.32.9 works, but 0.32.10 doesn't.
Just a note on driver versions of Redshift:
https://github.com/metabase/metabase/blob/master/modules/drivers/redshift/project.clj
- 0.33.5 to now (0.33.5.1) is `1.2.36.1060`
- 0.33.1 to 0.33.4 was `1.2.32.1056`
- 0.32.6 to 0.33.0 was `1.2.27.1051`
- 0.32.0 to 0.32.5 was `1.2.18.1036`

**Comment by GitsMcGee** [comment]
I take it it's safe to downgrade from 0.33.5.1 to 0.32.9? I'm always leery of going too far back and jacking up the metadata.

**Comment by flamber** [comment]
@GitsMcGee Always make a backup! Just in case :wink: Also do that before upgrades.
But if you're using non-H2 for metadata, then you should be able to downgrade between those two specific versions. If you're using H2, then you cannot downgrade from 0.33.5+ to earlier version, since the schema changed, so all queries will fail with a `does not match schema` error.

**Comment by patricktrainer** [comment]
> I can confirm @ifosch 's findings: downgrading to `v0.32.9` fixed it

I'll second that this worked for me as well.

**Comment by GitsMcGee** [comment]
I see that a new Redshift JDBC driver has been released (1.2.37.1061), but no mention of this issue in the notes: https://s3.amazonaws.com/redshift-downloads/drivers/jdbc/1.2.37.1061/Amazon+Redshift+JDBC+Release+Notes.pdf

**Comment by reidmweber** [comment]
I had tried switching the database driver to postgres (I'm using Panoply), which seemed to fix the connection issue but then I was getting this error for a bunch of queries. 
ERROR: Specified types or functions (one per INFO message) not supported on Redshift tables.
I also tried to go back to 32.9 but I had issues with most queries having errors so I went back to 33.6 for now.

**Comment by flamber** [comment]
Seems like it might have something to do with TLS1.3 ([search on Github](https://github.com/search?q=extension+%285%29+should+not+be+presented+in+certificate_request&type=Issues)).
I don't know if that can be controlled on EBS, as a workaround, to force it to TLS1.2.
Metabase changed from OpenJDK 8 to AdoptOpenJDK 11 in Docker images on 0.32.10, which is why 0.32.9 works, since it probably doesn't support TLS1.3 (I guess).

**Comment by GitsMcGee** [comment]
I'm in the same boat as @reidmweber, completely dead in the water after trying all of those things. I wonder what the ratio is of people having this problem who are using Redshift directly vs a go-between such as Panoply. I am using Panoply, and it sounds like a few people on this thread are as well. 

On that note, I was able to reproduce this same error using a different database tool that I set up on my local machine, I reached out to Panoply and they instructed me to switch the JDBC driver to 1.2.1.1001, which worked like a charm. As a stop-gap, would it be possible to publish a special build of 0.33.6 with the 1.2.1.1001 Redshift driver in it?

**Comment by sharedphysics** [comment]
Jumping in to say that it suddenly stopped working for me two days ago as well.
 
We are set up on AWS Elastic Beanstalk, running 0.33.2. I upgraded to 0.33.6 with the same issue. 

I am able to connect fine to google analytics, internal sample data, but having a problem connecting to Panoply (AWS Redshift). I tried connecting to Panoply with a Postgres driver and it gave me a different error: `SSL error: extension (5) should not be presented in certificate_request`

@GitsMcGee - How did you switch over to 1.2.1.1001 JDBC driver to fix?

**Comment by salsakran** [comment]
Hey all - We've reached out to the redshift team to see if we can figure out what's going on.

Sorry for the disruption =(

**Comment by GitsMcGee** [comment]
> @GitsMcGee - How did you switch over to 1.2.1.1001 JDBC driver to fix?

@sharedphysics I didn't. Not with Metabase. In an effort to try and get a deeper understanding of what was going on and what was to blame, I installed a different product (https://www.jetbrains.com/datagrip/) on a different machine (my Mac) to see if I could reproduce the same `SSL error: extension (5) should not be presented in certificate_request` error from that set up. I was able to reproduce it, and when I talked to their support they told me to change the driver version to `1.2.1.1001`, which in DataGrip you can do with a simple dropdown.

**Comment by psokolov-amzn** [comment]
extension (5) is status_request, see RFC3546. This smells like https://github.com/openssl/openssl/issues/10007

**Comment by ppp0** [comment]
Here a  comparison of SSL/TLS handshakes against `db.panoply.io:5439`

**TL;DR: no session established with TLSv1.3**

#### TLSv1

```
$ openssl s_client -tls1 -servername db.panoply.io -connect db.panoply.io:5439
CONNECTED(00000003)
write:errno=0
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 0 bytes and written 126 bytes
Verification: OK
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1
    Cipher    : 0000
    Session-ID: 
    Session-ID-ctx: 
    Master-Key: 
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    Start Time: 1574413696
    Timeout   : 7200 (sec)
    Verify return code: 0 (ok)
    Extended master secret: no
---
```

#### TLSv1.2

```
$ openssl s_client -tls1_2 -servername db.panoply.io -connect db.panoply.io:5439
CONNECTED(00000003)
write:errno=0
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 0 bytes and written 224 bytes
Verification: OK
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : TLSv1.2
    Cipher    : 0000
    Session-ID: 
    Session-ID-ctx: 
    Master-Key: 
    PSK identity: None
    PSK identity hint: None
    SRP username: None
    Start Time: 1574413716
    Timeout   : 7200 (sec)
    Verify return code: 0 (ok)
    Extended master secret: no
---
```

#### TLSv1.3
```
$ openssl s_client -tls1_3 -servername db.panoply.io -connect db.panoply.io:5439
CONNECTED(00000003)
write:errno=0
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 0 bytes and written 239 bytes
Verification: OK
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
Early data was not sent
Verify return code: 0 (ok)
---
```

**Comment by solokacher** [comment]
Hi, @ppp0 , pardon my lack of knowledge of panoply here. Redshift uses Postgres protocol. So you cannot just use openssl s_client to connect to it, as it has some preceding plaintext communications before the TLS handshake.

I see you are testing connections to db.panoply.io, but not the redshift cluster itself. I think unless panoply sneakily manipulates the postgres protocol bytes, you are only testing panoply, not Redshift.

**Comment by ppp0** [comment]
>you are only testing panoply, not Redshift.
I agree. I would guess, Panoply hosts its Redshift installation at Amazon's and that the address I tested against (db.panoply.io) is a proxy of some kind. I only wanted to show the difference when using different TLS protocol version

>..useless panoply..
You seem to have used it lots, haven't you

**Comment by solokacher** [comment]
Hi @ppp0, please copy&paste carefully... I said "UNLESS panoply", not "useless panoply". And, I don't use panoply at all. Not sure how you can tell "seem to have used it lots" here.

Postgres protocol requires client to send a plaintext packet saying "I want to do TLS handshake", and server replies a single byte 'S', and then TLS handshake happens. Unlike HTTPS and many other TLS-based protocol, TLS handshake happens right after TCP connection establishment.

If you are still in doubt, you can quickly install a postgres server on your machine (there are plenty of prebuilt binaries for all platforms) and test if you can use openssl s_client to connect to it.

- Kevin

**Comment by paddoq** [comment]
We are on Metabase version v0.33.0 , and the connection to panoply stop working last week . We switched from redshift driver to postgres driver and it started working again, but with issues, new fields and table are not coming through. 

Will appreciate any update on this.

Thank you.

**Comment by salsakran** [comment]
@patricktrainer @GitsMcGee @sharedphysics @reidmweber 

Is anyone experiencing this but not using panoply? 

Can someone who is having an issue email me directly (sameer @ metabase) ? We'd like to connect you to the redshift team who are investigating.

**Comment by paddoq** [comment]
Hi Sameer i emailed - in case you missed it or hit spam, you can reach me directly on 
patrick @ dinesafe  .org

or pass on to redshift team . thanks

**Comment by sharedphysics** [comment]
Hi Sameer, emailed. 
I have a second Metabase instance running 32.10 on AWS and also having the same problem. It is reading from a AWS-hosted Postgres data warehouse -- so _not_ Panoply.

**Comment by avivklas** [comment]
Hey guys, 

I'm an engineer in Panoply in charge of this issue. After investigating this we've found that the issue had been caused by a difference in the implementation of the TLS 1.3 protocol between Java and Go.
We upgraded our SQL Gateway server which is written in Go to use TLS 1.3 which caused an error for Java clients trying to connect to it.
The issue is already known by the Go team which claims that Java didn't respect the protocol spec exactly as it should, but they are willing to work around it in the next version of Go.

Today we reverted this upgrade in order to allow Java clients connecting to us back again so you can already try Metabase with Redshift JDBC driver and see if it works for you.

Please let me know if you experiencing any other issues with connecting Metabase to Panoply using Redshift driver or contact our support with any other issues.

**Comment by sharedphysics** [comment]
Hi @avivklas - Confirming it is working for me now. Thanks!

**Comment by ppp0** [comment]
@avivklas same here. Works. Big thank you

**Comment by paddoq** [comment]
Hi @avivklas Thanks very much - working now. 

ls panoply able to put some tests or measures in place to reduce risk of an outage like this again? For example before rolling out an upgrade to production, ensure that the major connectors work correctly. I understand you mention it was an issue from another team, but all the same, it means an outage for an end user like me, and it would be helpful to know that some controls are in place to reduce the risk of something like this happening again. 

Thank you.

**Comment by GitsMcGee** [comment]
@avivklas Working for me as well.

**Comment by salsakran** [comment]
thanks @avivklas closing
